File: LC/49-Group_anagrams/groupAnagrams.c
/**
 * Return an array of arrays of size *returnSize.
 * The sizes of the arrays are returned as *returnColumnSizes array.
 * Note: Both returned array and *columnSizes array must be malloced, assume caller calls free().
 */

#include <unistd.h>
#include <stdlib.h>

char	***groupAnagrams(char **strs, int strsSize,
						int *returnSize, int **returnColumnSizes)
{
	char ***ret;

	ret = malloc(sizeof(char ***));
	
	return (ret);
}
// {{"asc", "csa"}, {"abd"}} -> returnSize = 2
// {{"asc", "csa"}, {"abd"}} -> returnColsize = {2, 1};


int	main(void)
{
	int	strcnt = 6;
	char strs[7][4] = {"eat\0", "tea\0", "tan\0", "ate\0", "nat\0", "bat\0", 0};
	int set_count;
	int set_lens[6] = {0, 0, 0, 0, 0, 0};
	char ***ret;

	ret = groupAnagrams((char **)strs, strcnt, &set_count, (int**)set_lens);
	free(ret);

	return (0);
	
}

File: LC/49-Group_anagrams/group_anagrams.py
from typing import List
from collections import Counter

class Solution:
	def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
		ret = [];
		while strs:
			cmp, *strs = strs;
			idx_arr = get_anagrams(strs, cmp);
			ret += [[cmp ] + list(map(lambda i: strs[i], idx_arr))];
			strs = [v for i, v in enumerate(strs) if i not in idx_arr];
		return ret;

def	get_anagrams(strs, cmp):
	cmp = Counter(cmp);
	return [i for i, v in enumerate(list(map(Counter, strs))) if v == cmp];

if __name__ == "__main__":
	strs = "eat,tea,tan,ate,nat,bat".split(",");
	strs.append("");
	strs.append("");
	strs.append("");
	expect = [["bat"], ["nat", "tan"], ["ate", "eat", "tea"]];
	a_sol = Solution();
	ret = a_sol.groupAnagrams(strs);
	print(ret);
	strs = "dddddt,tttttd".split(",");
	ret = a_sol.groupAnagrams(strs);
	print(ret);
